src/main/java/net/dries007/tfc/common/blocks/devices/CharcoalForgeBlock.java

```diff
@@ -0,0 +1,170 @@
+            .match(new BlockPos(-1, 0, 0), isValidSide)
+            .match(new BlockPos(0, 0, 1), isValidSide)
+            .match(new BlockPos(0, 0, -1), isValidSide)
+            .match(new BlockPos(0, -1, 0), isValidSide);
+    }
+
+    public static boolean isValid(IWorld world, BlockPos pos)
+    {
+        return FORGE_MULTIBLOCK.test(world, pos);
+    }
+
+    public CharcoalForgeBlock(ForgeBlockProperties properties)
+    {
+        super(properties);
+        registerDefaultState(getStateDefinition().any().setValue(HEAT, 0));
+    }
+
+    @OnlyIn(Dist.CLIENT)
+    @Override
+    public void animateTick(BlockState state, World world, BlockPos pos, Random rand)
+    {
+        if (state.getValue(HEAT) == 0) return;
+        double x = pos.getX() + 0.5D;
+        double y = pos.getY() + 0.875D;
+        double z = pos.getZ() + 0.5D;
+
+        if (rand.nextInt(10) == 0)
+        {
+            world.playLocalSound(x, y, z, SoundEvents.FIRE_AMBIENT, SoundCategory.BLOCKS, 0.5F + rand.nextFloat(), rand.nextFloat() * 0.7F + 0.6F, false);
+        }
+        for (int i = 0; i < 1 + rand.nextInt(2); i++)
+        {
+            world.addAlwaysVisibleParticle(ParticleTypes.LARGE_SMOKE, x + Helpers.fastGaussian(rand), y + rand.nextDouble(), z + Helpers.fastGaussian(rand), 0, 0.07D, 0);
+        }
+        for (int i = 0; i < rand.nextInt(3); i++)
+        {
+            world.addParticle(ParticleTypes.SMOKE, x + Helpers.fastGaussian(rand), y + rand.nextDouble(), z + Helpers.fastGaussian(rand), 0, 0.005D, 0);
+        }
+        if (rand.nextInt(8) == 1)
+        {
+            world.addParticle(ParticleTypes.LAVA, x + Helpers.fastGaussian(rand), y + rand.nextDouble(), z + Helpers.fastGaussian(rand), 0, 0.005D, 0);
+        }
+    }
+
+    @Override
+    public void stepOn(World world, BlockPos pos, Entity entity)
+    {
+        if (!entity.fireImmune() && entity instanceof LivingEntity && !EnchantmentHelper.hasFrostWalker((LivingEntity) entity) && world.getBlockState(pos).getValue(HEAT) > 0)
+        {
+            entity.hurt(DamageSource.HOT_FLOOR, 1.0F);
+        }
+        super.stepOn(world, pos, entity);
+    }
+
+    @Override
+    protected void createBlockStateDefinition(StateContainer.Builder<Block, BlockState> builder)
+    {
+        super.createBlockStateDefinition(builder);
+        builder.add(HEAT);
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public BlockState updateShape(BlockState state, Direction facing, BlockState facingState, IWorld world, BlockPos currentPos, BlockPos facingPos)
+    {
+        return state.getValue(HEAT) > 0 && !isValid(world, currentPos) ? state.setValue(HEAT, 0) : state;
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public ActionResultType use(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockRayTraceResult result)
+    {
+        if (player instanceof ServerPlayerEntity && !player.isShiftKeyDown())
```

Comment:
the `ServerPlayerEntity` should be moved inside. effectively:
```java
if (/* conditions that can be checked on both sides, and may fail */) {
    if (/* server side */) {
        // server side specific actions
    }
    return SUCCESS; // return success on both sides
}
return PASS; // return pass or fail if something didn't, or couldn't happen.
```

I'm also fairly convinced you don't need the shift check at all because shifting bypasses the block interaction by default. Meaning this can just be a TE -> open gui on server -> always return success chain.